servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java

```diff
@@ -71,74 +70,58 @@ public static ConcurrentSubscription wrap(Subscription subscription) {
-        if (currentThread == subscriptionLockOwner) {
-            subscriptionRequestQueueUpdater.accumulateAndGet(this, n,
-                    FlowControlUtils::addWithOverflowProtectionIfNotNegative);
-            return;
-        }
-        do {
-            if (!subscriptionLockOwnerUpdater.compareAndSet(this, null, currentThread)) {
-                // It is possible that we picked up a negative value from the queue on the previous iteration because
-                // we have been cancelled in another thread, and in this case we don't want to increment the queue and
-                // instead we just set to MIN_VALUE again and try to re-acquire the lock in case we raced again.
-                if (n < 0) {
-                    subscriptionRequestQueueUpdater.set(this, Long.MIN_VALUE);
-                } else {
-                    subscriptionRequestQueueUpdater.accumulateAndGet(this, n,
-                            FlowControlUtils::addWithOverflowProtectionIfNotNegative);
-                }
-                if (!subscriptionLockOwnerUpdater.compareAndSet(this, null, currentThread)) {
-                    return;
-                }
-                // We previously added our n contribution to the queue, but now that we have acquired the lock
-                // we are responsible for draining the queue.
-                n = subscriptionRequestQueueUpdater.getAndSet(this, 0);
-                if (n == 0) {
-                    // It is possible that the previous consumer has released the lock, and drained the queue before we
-                    // acquired the lock and drained the queue. This means we have acquired the lock, but the queue has
-                    // already been drained to 0. We should release the lock, try to drain the queue again, and then
-                    // loop to acquire the lock if there are elements to drain.
-                    subscriptionLockOwner = null;
-                    n = subscriptionRequestQueueUpdater.getAndSet(this, 0);
-                    if (n == 0) {
-                        return;
-                    } else {
-                        continue;
-                    }
+            pendingDemand = mapInvalidRequestN(n);
+        } else {
+            for (;;) {
+                final long prevPendingDemand = pendingDemand;
+                if (prevPendingDemand < 0 ||
+                        pendingDemandUpdater.compareAndSet(this, prevPendingDemand,
+                                prevPendingDemand + min(Long.MAX_VALUE - prevPendingDemand, n))) {
+                    break;
                 }
             }
-            if (n < 0) {
-                subscription.cancel();
-                return; // Don't set subscriptionLockOwner = 0 ... we don't want to request any more!
+        }
+        Throwable delayedCause = null;
+        boolean tryAcquire;
+        do {
+            final long acquireId = tryAcquireReentrantLock(subscriptionLockUpdater, this);
```

Comment:
good point, I also considered this. the re-entrant lock isn't strictly necessary for correctness as we could use the non-reentrant lock which also requires re-acquire semantics. for re-entry scenarios this approach saves a CAS on each re-entrant operation and trades it for a `Thread.getId()` (which is thread local state). Current approach preserves the previous re-entrant semantics in this class to minimize behavior difference and also shows perf improvement:

```
# Reentrant lock (this PR)
Benchmark                                     Mode  Cnt         Score        Error  Units
ConcurrentSubscriptionBenchmark.noReentrant  thrpt    5  47753164.225 ± 850903.330  ops/s
ConcurrentSubscriptionBenchmark.reentrant    thrpt    5    714751.803 ±   8155.984  ops/s

Benchmark                                     Mode  Cnt         Score         Error  Units
ConcurrentSubscriptionBenchmark.noReentrant  thrpt    5  48522938.411 ± 1065238.470  ops/s
ConcurrentSubscriptionBenchmark.reentrant    thrpt    5    723176.975 ±   11730.345  ops/s

Benchmark                                     Mode  Cnt         Score        Error  Units
ConcurrentSubscriptionBenchmark.noReentrant  thrpt    5  48460738.326 ± 918765.253  ops/s
ConcurrentSubscriptionBenchmark.reentrant    thrpt    5    718663.755 ±  22510.103  ops/s

# Non-reentrant (this PR modified to use tryAcquireLock)
Benchmark                                     Mode  Cnt         Score         Error  Units
ConcurrentSubscriptionBenchmark.noReentrant  thrpt    5  46236169.698 ± 1304606.373  ops/s
ConcurrentSubscriptionBenchmark.reentrant    thrpt    5    379186.656 ±    7327.488  ops/s

Benchmark                                     Mode  Cnt         Score        Error  Units
ConcurrentSubscriptionBenchmark.noReentrant  thrpt    5  46554069.756 ± 770443.754  ops/s
ConcurrentSubscriptionBenchmark.reentrant    thrpt    5    374243.264 ±   4981.803  ops/s

Benchmark                                     Mode  Cnt         Score        Error  Units
ConcurrentSubscriptionBenchmark.noReentrant  thrpt    5  46420860.395 ± 906316.830  ops/s
ConcurrentSubscriptionBenchmark.reentrant    thrpt    5    379715.245 ±   7939.865  ops/s
```

Reply:
> Current approach preserves the previous re-entrant semantics in this class to minimize behavior difference

We are modifying an internal class in a way that adheres to the specifications. If we do not depend on this behavior (which we shouldn't), I don't think there is anything wrong in simplification. Allowing re-entrancy here adds more complexity in this code and additionally in `ConcurrentUtils`. Removing re-entrancy reduces thread-stack which is a net win. 

I don't think the small perf win trumps the other benefits.

Reply:
The benchmark was helpful :) detected regression from master on the no-reentrant case. master skips the atomic long accumulator if we are able to acquire the lock, but trades some concurrency. I added a fast path when no concurrency is detected and we now gain in both cases (re-entrant and non-reentrant) which is assumed to be more common, and even if there is concurrency the expected cost is higher.

```
# Re-entrant with fast path
Benchmark                                     Mode  Cnt         Score        Error  Units
ConcurrentSubscriptionBenchmark.noReentrant  thrpt    5  74805386.276 ± 746665.173  ops/s
ConcurrentSubscriptionBenchmark.reentrant    thrpt    5   1895818.132 ±  31661.798  ops/s

Benchmark                                     Mode  Cnt         Score        Error  Units
ConcurrentSubscriptionBenchmark.noReentrant  thrpt    5  74602546.497 ± 972834.986  ops/s
ConcurrentSubscriptionBenchmark.reentrant    thrpt    5   2035494.125 ±  28986.077  ops/s

Benchmark                                     Mode  Cnt         Score        Error  Units
ConcurrentSubscriptionBenchmark.noReentrant  thrpt    5  74724470.463 ± 809942.225  ops/s
ConcurrentSubscriptionBenchmark.reentrant    thrpt    5   1864583.823 ±  48146.217  ops/s

# master
Benchmark                                     Mode  Cnt         Score        Error  Units
ConcurrentSubscriptionBenchmark.noReentrant  thrpt    5  59009939.298 ± 913522.147  ops/s
ConcurrentSubscriptionBenchmark.reentrant    thrpt    5    449874.916 ±  10856.180  ops/s

Benchmark                                     Mode  Cnt         Score        Error  Units
ConcurrentSubscriptionBenchmark.noReentrant  thrpt    5  58815916.826 ± 963463.757  ops/s
ConcurrentSubscriptionBenchmark.reentrant    thrpt    5    447968.057 ±   6171.099  ops/s

Benchmark                                     Mode  Cnt         Score         Error  Units
ConcurrentSubscriptionBenchmark.noReentrant  thrpt    5  58929994.344 ± 1208720.403  ops/s
ConcurrentSubscriptionBenchmark.reentrant    thrpt    5    446883.247 ±   15198.500  ops/s
```

Reply:
Here are the numbers for the `tryAcquireLock` version with fast path. The noReentrant is similar as expected now but the re-entrant case is much worse (only ~18% of the perf). In order to avoid regression I think it makes sense to move forward with the approach in this PR (as opposed to using `tryAcquireLock`).

```
# Non-reentrant (this PR with fast path, modified to use tryAcquireLock)
Benchmark                                     Mode  Cnt         Score         Error  Units
ConcurrentSubscriptionBenchmark.noReentrant  thrpt    5  75341058.045 ± 1011207.561  ops/s
ConcurrentSubscriptionBenchmark.reentrant    thrpt    5    337948.564 ±    5698.630  ops/s
```

Reply:
From my perspective, breaking recursion here is better than allowing it as it cuts down the thread stack. 

I would suggest avoiding the complexity of adding reentrant lock semantics in `ConcurrentUtils`. I am also not completely convinced about the approach for the reentrant lock but we can discuss that in https://github.com/apple/servicetalk/pull/1015#discussion_r417523512

Reply:
It isn't the responsibility of this class to limit recursion/stack depth. Limiting stack depth is shown in the benchmarks to reduce performance.